### DISCUSSION TITLE
Interfaces

### DISCUSSION THREAD

**Discussion by kotyara85** [rabbitmq/amqp091-go#53] [Ideas]
I’d love to see go interfaces as a part of this package. It does make everyone’s like miserable to write unit tests when you have integration with amqp. 
I can take care or that.

**Comment by mihaitodor** [comment]
I had a conversation with the original author on this topic back in 2017: https://github.com/streadway/amqp/issues/164#issuecomment-271523006 Not sure if it's still relevant, but thought I'd point it out.

**Comment by Zerpet** [comment]
The conversation in https://github.com/streadway/amqp/issues/164#issuecomment-271523006 seems relevant to me. Making this package return interfaces will be a breaking change; we are happy to consider reasonable breaking changes, however, in this case, the solution proposed by streadway is valid, and I don't think we should introduce a breaking change here.

**Comment by kewinbrand** [comment]
how you guys test your code being unable to mock?

**Comment by mihaitodor** [comment]
@kewinbrand I'd have to see specific examples, but you can define a bunch of wrappers in your code as described [here](https://stackoverflow.com/questions/64501096/how-to-mock-functions-from-libraries-like-amqp-dial). That might get a bit verbose to maintain and the [`github.com/NeowayLabs/wabbit`](https://github.com/NeowayLabs/wabbit) library seems to contain all this plumbing already, so you can import it and use it throughout your code. They adjusted it to delegate to `rabbitmq/amqp091-go` instead of `streadway/amqp` recently, so there shouldn't be any issues.

**Comment by noredine69** [comment]
Hi folks, and @kewinbrand
I've created a mock, using read.go, write.go, spec_091.go, types.go.
In my code, a tcp server is created, like in client_test.go, and I wrote the expected message received and send the corresponding response. 
In the client_test.go file, it's clearly done in lines 327 and 328,
the server is expecting a connectionOpen message. It will send a connectionOpenOk as soon as it receives the request.
In the go routine (line 319 to 331 of the file client_test.go), you have to write the  protocol corresponding to the code contained in the main thread (line 333).
The hard part is to maintained the expected messages (in the go routine) when the main code is modified.
The ugly part is that I've to duplicate many files from source code

**Comment by DanielePalaia** [comment]
Hi all, this one seems more a discussion at this point, let's convert it to a discussion while we decide the best way to go.

**Comment by mihaitodor** [comment]
Besides [github.com/NeowayLabs/wabbit](https://github.com/NeowayLabs/wabbit), I noticed there's also https://github.com/wagslane/go-rabbitmq, which is described as a "wrapper of [rabbitmq/amqp091-go](https://github.com/rabbitmq/amqp091-go) that provides reconnection logic and sane defaults". Might be worth looking at before making any changes.
